runtime/block_reduction.cu

```diff
@@ -48,29 +48,103 @@ __device__ void blockReduce(
-  unsigned int smem_offset = reduction_idx * reduction_size + reduction_tid;
+  unsigned int smem_offset = threadIdx.x + threadIdx.y * blockDim.x +
+      threadIdx.z * blockDim.x * blockDim.y;
+
+  // The peer stride represents the distance between the current element and its
+  // nearest reduction peer. It depends on the reduction dimension. A reduction
+  // peer refers to elements that belong to the same reduction segment. For
+  // example, if the reduction is across TIDy, all the elements in the same
+  // column (with the same TIDx) are considered peers of each other. The
+  // distance between an element and its nearest peer is blockDim.x.
+  constexpr int num_redu_dims = (int)X_REDUCE + (int)Y_REDUCE + (int)Z_REDUCE;
+  constexpr bool xz_reduce = (num_redu_dims == 2 && !Y_REDUCE);
+  // reduction in 3 dimensions, XYZ, stride is 1
+  unsigned int peer_stride = 1;
+  if (num_redu_dims == 1) {
+    // Reduction only in 1 dimension, X or Y or Z
+    // e.g. inner or outer reduction
+    // If X_REDUCE, reducing in neighbor cols in smem, peer_stride is 1
+    // If Y_REDUCE, reducing in neighbor rows in smem, peer_stride is blockDim.x
+    // If Z_REDUCE, reducing in neighbor planes in smem, peer_stride is
+    // blockDim.x * blockDim.y
+    peer_stride = X_REDUCE ? 1
+        : Y_REDUCE         ? blockDim.x
+                           : blockDim.x * blockDim.y;
+  } else if (num_redu_dims == 2) {
+    // Reduction in 2 dimensions, only one dimension is not reduced, !X, !Y, !Z
+    // If !Z_REDUCE, merge XY, reducing neighbor cols, peer_stride is 1
+    // If !X_REDUCE, merge ZY, reducing neighbor rows, peer_stride is blockDim.x
+    // If !Y_REDUCE, if blockDim.y == 1, merge XZ, peer_stride is 1.
+    // otherwise, needs carefully calculate offset to the reduction peer:
+    // (1) redu_offset = reduction_tid + tree_fold_factor
+    // (2) idz = redu_offset / blockDim.x
+    // (3) idx = redu_offset % blockDim.x
+    // (4) smem_offset = idx + threadIdx.y * blockDim.x + idz * blockDim.x *
+    // blockDim.y
+    if (!Y_REDUCE) {
+      peer_stride = 1;
+    } else {
+      peer_stride = !Z_REDUCE ? 1 : blockDim.x;
+    }
+  }
 
   // Initialize shared memory
   if (read_pred) {
     shared_mem[smem_offset] = inp_val;
   } else {
     shared_mem[smem_offset] = init_val;
   }
-
   block_sync::sync<Aligned>();
+
   // Reduce down to nearest power of 2 for the tree reduction:
   int np2 = 1 << (31 - __clz(reduction_size));
-
   if (reduction_tid < np2 && reduction_tid + np2 < reduction_size) {
-    reduction_op(shared_mem[smem_offset], shared_mem[smem_offset + np2]);
+    int peer_offset = smem_offset + np2 * peer_stride;
+    if constexpr (xz_reduce) {
+      if (blockDim.y > 1) {
+        int redu_offset = reduction_tid + np2;
+        int idz = redu_offset / blockDim.x;
+        int idx = redu_offset % blockDim.x;
+        peer_offset =
+            idx + threadIdx.y * blockDim.x + idz * blockDim.x * blockDim.y;
```

Comment:
We have `reduction_tid = threadIdx.z * blockDimx + threadIdx.x`.
and that gives us
`idz = threadIdx.z + (threadIdx.x + np2) / blockDimx`. This non-linear part I feel could lead to bank-conflict?

Does block reduction by itself go through `xz_reduce`? or is it only accessible through grid reduction?

Reply:
Actually, let me try to run some concrete number with this one. :expressionless:

Reply:
This block reduction supports reduce in X/Y/Z dims, so there are 2^3 = 8 configurations. Assume at least 1 reduction dimension, there are 8 - 1 = 7 different actual configurations. It can do `xz_reduce` by itself, the newly added test `BlockReduction3D` is testing this `xz_reduce`.

I did a solid example for your reference:
![image](https://github.com/NVIDIA/Fuser/assets/116412316/b63f6724-f81e-418c-af8b-a1a7cf02d117)
Color represents different warps, at each stage, different threads within the same warp are accesssing different banks. For example, in the 1st loop,  in the 1st transaction, T0 is loading T0, T1 is loading T1, ..., Tn is loading Tn (where n<128), then in the 2nd transaction, T0 is loading T128, T1 is loading T129, ..., Tn is loading Tn+128 (where n< 128). In the last loop, T0 is loading T0 & T1, T8 is loading T8 & T9, 4 active threads per warp, each thread is loading a different bank. The issue is we have too many idle threads in the last several loops instead of bank conflicts.

Reply:
Double checked with ncu using test `BlockReduction3D` with Z,Y,X= [4,8,8], there is no bank conflict.
![image](https://github.com/NVIDIA/Fuser/assets/116412316/0de65cdf-c69e-4ae8-8bc8-508fc103b97a)

Reply:
Thanks for walking through the example.

> idx = (threadIdx.x + np2) % blockDim.x
> peer_offset = idx + threadIdx.y * blockDim.x + idz * blockDim.x * blockDim.y;
> shared_mem[peer_offset]

You are right about the bank conflict part. mod by blockDim.x should be safe, since there's a single tidx in the indexing part... Sorry about the noise.